src/main/haskell/kore/src/Kore/Unification/Substitution.hs

```diff
@@ -0,0 +1,110 @@
+{-|
+Module      : Kore.Unification.Substitution
+Description : The Substitution type.
+Copyright   : (c) Runtime Verification, 2018
+License     : NCSA
+Maintainer  : vladimir.ciobanu@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+
+module Kore.Unification.Substitution
+    ( Substitution
+    , unwrap
+    , wrap
+    , modify
+    , mapSubstitutionVariables
+    , isNormalized
+    , null
+    , variables
+    , unsafeWrap
+    ) where
+
+import Control.DeepSeq
+       ( NFData )
+import GHC.Generics
+       ( Generic )
+import Prelude hiding
+       ( null )
+
+import Kore.AST.PureML
+import Kore.Step.Pattern
+       ( StepPattern )
+
+data Substitution level variable
+    = Substitution ![(variable level, StepPattern level variable)]
```

Comment:
My main problem with this is that we are sometimes representing unnormalized substitutions as predicates, e.g. when we unify `concat(elem(x, y), z))` with a map. By default I would have expected to have a (un)normalized substitution-predicate, but this approach may also make sense. If that's the case, could you add some documentation (which you should have for all public stuff anyway) explaining, among others, the expected usage?